### PR TITLE
AK: Add initial support for runtime detection of SIMD CPU capabilities

### PR DESCRIPTION
**Pull request by DanShaders**
This allows to easily detect at runtime which vector extensions CPU actually supports in order to use the best version of the manually vectorized algorithm, regardless of the compile options.

Unfortunately, no uses with this PR. However, the code is a part of #16126 (and [used](https://github.com/DanShaders/serenity/blob/1bd04fe15d7ce4c1fd14c12d2c252ed6ea7b3d4f/AK/BigIntBase.cpp#L639) there) and it looks as if @Zaggy1024 might want to use it too (see discussion in #multimedia)

### CODE REVIEW

**Comment by Hendiadyoin1** [comment]
Iirc LibC/DynamicLoader uses runtime feature detection, you may want to have a look on that

**Comment by Hendiadyoin1** [comment]
Never mind, that just detects TCG, and never touches more than sse2:
https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibC/arch/x86_64/memset.cpp

File: AK/SIMD.h [hunk 1]
```diff
@@ -10,6 +10,88 @@
...
+    static u32 cached = static_cast<u32>(NONE);
+    if (has_cached)
+        return cached;
+
+    __builtin_cpu_init();
+    if (__builtin_cpu_supports("avx2"))
+        cached |= static_cast<u32>(AVX2);
+
+    has_cached = true;
+    return cached;
```
Comment (BertalanD) [inline AK/SIMD.h:47]:
```suggestion
    static u32 s_cached = [] {
        if (__builtin_cpu_supports("avx2"))
            return static_cast<u32>(AVX2);
        return static_cast<u32>(NONE);
    } ();
    return s_cached;
```

This also suffices if you don't mind the load from `s_cached` being thread-safe/atomic. ([Godbolt](https://gcc.godbolt.org/z/K8qqYszMa))

Or the caching might not even be necessary. `__builtin_cpu_init()` is called as a global constructor anyway, so unless you're calling SIMD code before `main()`, calling the function here is redundant. And the check compiles down to very simple code:
```asm
main:
        mov     eax, DWORD PTR __cpu_model[rip+12]
        shr     eax, 10
        and     eax, 1
        ret
```

([Godbolt](https://gcc.godbolt.org/z/87Tax1exe))

Comment (DanShaders) [reply AK/SIMD.h]: Oh, this is much simpler :) I somehow skipped the note about global constructor while reading the documentation.

**Comment by nico** [comment]
We usually do processor capability detection on the kernel (https://github.com/SerenityOS/serenity/blob/7f52f45e9d7704b684714201af138974adfcfcfe/Kernel/Arch/x86/CPUID.h#L113) and expose that via sysctl or getconf or something like that. Could that work here?

**Comment by DanShaders** [comment]
I guess this would work, although I'm not entirely sure why we need to invent a special, non-portable syscall parameter to get information, which we already automatically have, thanks to the compiler.

**Comment by BertalanD** [comment]
I think it's easier and nicer to use `__builtin_cpu_supports()` than to make extra syscalls and whatnot to detect CPU features. The support data structures are initialized for it by libgcc's static constructors anyway, so why not just use them?

AArch64 *will* be an exception though, as CPU feature flags are not exposed to EL0, so feature detection *has to* go through the kernel there.

**Review by linusg — CHANGES_REQUESTED** [review]
We don't add dead code, add this in a PR that also uses it.

**Comment by DanShaders** [comment]
Sure

**Pull request closed without merging**